Update regex literal delimiters

hamishknight opened the pull request:
Update the lexing code for the replacement of the `'/.../'` and `'|...|'` delimiters with `#/.../#` and `#|...|#` respectively, in addition to allowing the `re'...'` delimiter (implemented in https://github.com/apple/swift-experimental-string-processing/pull/167).

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

rxwei commented on test/StringProcessing/Sema/regex_literal_type_inference.swift:
```diff
@@ -1,48 +1,48 @@
 //                 ^ 2
 //                   ^ 3
-let _: Regex<(Substring, [Substring]?, [Substring?]?, [[Substring]?]?)>.Type = type(of: r7)
+let _: Regex<(Substring, Substring??, Substring???, Substring????)>.Type = type(of: r7)
```
Looks.... correct????

rintaro commented on lib/Parse/Lexer.cpp:
```diff
@@ -2404,6 +2401,11 @@ void Lexer::lexImpl() {
+  // If we have experimental string processing enabled, try to lex a regex
+  // literal.
+  if (tryLexRegexLiteral(TokStart))
+    return;
```
Doing this for every token might cause noticeable performance regression. Can we move this to
`case '#':` and `case 'r'`?

hamishknight replied:
Fair enough, how does this look?

hamishknight committed:
Update regex literal delimiters

Update the lexing code for the replacement of the
`'/.../'` and `'|...|'` delimiters with `#/.../#`
and `#|...|#` respectively, in addition to
allowing the `re'...'` delimiter.

hamishknight commented:
@swift-ci please test

rxwei commented:
Currently pending an integration workflow change (swift-experimental-string-processing#170).

rxwei commented:
@swift-ci please clean test

rxwei commented:
We've decided to use [dev/7](https://github.com/apple/swift-experimental-string-processing/releases/tag/dev%2F7) for this round of integration.

rxwei commented:
apple/swift-experimental-string-processing#186
@swift-ci please test

rxwei commented:
@swift-ci please test

rxwei commented:
Ok, this PR should be ready for @shahmishal to merge.

rxwei commented:
@swift-ci please test macOS
